### ISSUE TITLE
Support for condition variables

### ISSUE THREAD

**Issue by AndreLouisCaron** [gevent/gevent#1326]
Hi there!

I have a few programs where I use condition variables and while experimenting with gevent, I noticed that it has implementation for semaphores and events, but not for condition variables.  Issue #859 clearly states that the implementation is not necessary (it's implicitly monkey patched), but it feels odd to explicitly import `threading.Condition` in gevent-based code.  Also, I generally don't monkey-patch my gevent-based programs, so I cannot really count on this even if it does work.

Anyways, I've found myself implementing my own `Condition` class to pair with `gevent.lock.Lock` and I was wondering if it could be included in `gevent`.  Any interest for this?

Thanks,

André

**Comment by jamadden** [comment]
I think (without much evidence, granted) that most use of gevent is via monkey-patching, as that's one of its main advantages (it makes standard code "just work"). gevent tries to implement the minimal primitives and then allow the standard library to build on them; we use the standard library implementation of a condition and we run the standard library tests of a condition.

**Comment by AndreLouisCaron** [comment]
Yeah, I get that.  However, I prefer explicitly tying my code to gevent to monkey-patching (for new code, at least).  In those programs, it feels really strange to import `threading.Condition` and then monkey patch just to get `threading.Condition`  to work as desired.  I know this works, but it just feels so odd.

**Comment by jamadden** [comment]
It depends on the exact needs, but for gevent-aware code very often gevent provides other primitives that can take the place of a Condition to more succinctly express what's going on (and take advantage of the nature of cooperative tasking to make certain assumptions). These can be as simple as a list of waiting greenlets that get switched to, or it may be a `gevent.event.Event`, or you may use a work queue.

**Comment by AndreLouisCaron** [comment]
Yeah, I took a good look at other ways of writing this code.  So far, I've found:

* `gevent.event.Event`: this has the downside of not being able to notify a single waiter.
* `gevent.lock.Semaphore`: this has the downside of requiring the notifier to know how many waiters to resume.  At best, this is just syntactically heavy.  At worst, you have multiple `Condition` variables for a single `Lock` and the code is a maintenance nightmare.

> These can be as simple as a list of waiting greenlets that get switched to

I'm not sure what you're suggesting here.  Can you tell me more?

**Comment by jamadden** [comment]
>> These can be as simple as a list of waiting greenlets that get switched to

> I'm not sure what you're suggesting here. Can you tell me more?

Something like this (untested):

```python
waiting_greenlets = []

# To 'park'
waiting_greenlets.append(getcurrent())

# notify one async
get_hub().loop.run_callback(waiting_greenlets.pop().switch)

# notify all async
while waiting_greenlets:
    # The above
    ...

# notify one sync
waiting_greenlets.pop().switch()
```

The `Waiter` and `MultipleWaiter` primitives essentially do this.

**Comment by AndreLouisCaron** [comment]
That seems like an interesting API for implementing a `Condition` class -- likely much better that my initial approach with an `greenlet.event.Event` object per waiter :-)

If I put that in a `Condition` class, would you consider adding it to `gevent`?  ;-)
